feat(STONEINTG-1266): make data-acceptable-bundles repos public update-trusted-tasks

kasemAlem opened the pull request:
**Summary** 
Automatically makes data-acceptable-bundles repositories public after update-trusted-tasks runs.
**Changes**
1. New task: prepare-data-acceptable-bundles
 - Runs after update-trusted-tasks
 - Extracts repository URLs from the snapshot and transforms them to data-acceptable-bundles format (e.g.,      quay.io/org/repo → quay.io/org/data-acceptable-bundles)
 - Adds these repositories as new components in the snapshot with public: true
 - Outputs a modified snapshot as a trusted artifact

2. Refactored pipeline flow
 - Removed the early make-repo-public task that ran before update-trusted-tasks
 - make-repo-public now runs after prepare-data-acceptable-bundles
 - make-repo-public uses the modified snapshot to make all repositories public (original + data-acceptable-bundles)

[STONEINTG-1266](https://issues.redhat.com/browse/STONEINTG-1266)

seanconroy2021 commented on tasks/managed/prepare-data-acceptable-bundles/prepare-data-acceptable-bundles.yaml:
```diff
@@ -0,0 +1,131 @@
+        echo "$REPOSITORIES" | jq .
+        # Write result (already compact JSON)
+        # Use base64 encoding to preserve JSON structure when Tekton reads the result
+        echo "$REPOSITORIES" | base64 -w 0 > "$(results.repositories.path)"
```
prepare-data-acceptable-bundles writes the result as base64, but make-repo-public expects plain JSON.

kasemAlem replied:
fixed

seanconroy2021 commented on tasks/managed/make-repo-public/make-repo-public.yaml:
```diff
@@ -12,10 +12,16 @@ spec:
   params:
     - name: dataPath
```
These parameters are passed from the release controller, so they will never be empty (dataPath,snapshotPath). I dont think we need update the description here

kasemAlem replied:
fixed

seanconroy2021 commented on tasks/managed/prepare-data-acceptable-bundles/prepare-data-acceptable-bundles.yaml:
```diff
@@ -0,0 +1,131 @@
+      type: string
+      description: The name of the key in the ConfigMap that contains the CA bundle data
+      default: ca-bundle.crt
+  workspaces:
```
We use trusted artifacts (OCI registry) to share data between tasks. I think you will need to add use-trusted-artifact step action you can look at https://github.com/konflux-ci/release-service-catalog/blob/development/tasks/managed/update-trusted-tasks/update-trusted-tasks.yaml for an example.

kasemAlem replied:
agree, this is the right way
fixed 
thanks

mmalina commented on tasks/managed/make-repo-public/make-repo-public.yaml:
```diff
@@ -58,6 +64,13 @@ spec:
       type: string
       description: The name of the key in the ConfigMap that contains the CA bundle data
       default: ca-bundle.crt
+    - name: repositories
```
I don't think we want this. As Sean said, you need to use trusted artifacts. And so my suggestion how to achieve this is to have the new prepare-data* task alter the snapshot to include the new things that need to be made public. And then change the sourceDataArtifact parameter value for make-repo-public in this pipeline to use the result of that new task.

kasemAlem replied:
agree with Sean's comment , 
fixed it, 
thanks

mmalina commented on pipelines/managed/push-tekton-task-bundles-to-external-registry/push-tekton-task-bundles-to-external-registry.yaml:
```diff
@@ -424,6 +424,78 @@ spec:
+          value: "$(params.taskGitRevision)"
+      runAfter:
+        - update-trusted-tasks
+    - name: make-data-acceptable-bundles-repo-public
```
Why add a new copy of this task? This pipeline already has the make-repo-public task.

kasemAlem replied:
missed to remove it, 
forgot to fix it ,

kasemAlem replied:
fixed it

mmalina commented:
@kasemAlem can squash the commits into one? There's not need to keep updating the PR, but if you do want to update it, please use the Rebase option - that way no extra merge commit is added.

Also, can you make sure the commit message and PR description match the current version? It says it's adding a parameter to make-repo-public, but that's no longer the case.

kasemAlem committed:
feat(STONEINTG-1266): make data-acceptable-bundles repos public

- Add prepare-data-acceptable-bundles task to modify snapshot
- Update plr make-repo-public after prepare-data-acceptable-bundles
- Remove duplicate make-repo-public task instance
- Add tests for prepare-data-acceptable-bundles task

Signed-off-by: Kasem Alem <kalem@redhat.com>

mmalina commented on pipelines/managed/push-tekton-task-bundles-to-external-registry/push-tekton-task-bundles-to-external-registry.yaml:
```diff
@@ -391,9 +386,8 @@ spec:
-        - collect-registry-token-secret
         - push-snapshot
-    - name: update-trusted-tasks
+    - name: prepare-data-acceptable-bundles
```
This one should probably have `runAfter: update-trusted-tasks`, right? We like to have something set explicitly.

kasemAlem replied:
isn't what we have in lines 421, 422  on this file ? 

```
      runAfter:
        - update-trusted-tasks
```

mmalina replied:
oh, you're right, I didn't see it

mmalina commented on tasks/managed/prepare-data-acceptable-bundles/prepare-data-acceptable-bundles.yaml:
```diff
@@ -0,0 +1,238 @@
+          memory: 256Mi
+        requests:
+          memory: 256Mi
+          cpu: 150m
```
This probably doesn't even need this much resources, right? Nothing is happening in parallel and it only runs some basic utilities like jq or awk. So I'd say it shouldn't need more than the `use-trusted-artifact` step, wdyt?

kasemAlem replied:
you are totally right, fixed

mmalina commented on tasks/managed/prepare-data-acceptable-bundles/prepare-data-acceptable-bundles.yaml:
```diff
@@ -0,0 +1,238 @@
+        # Track unique data-acceptable-bundles repositories
+        REPOSITORIES='[]'
+
+        # Iterate through each component
```
This whole thing seems a bit inefficient. Because it parses the json over and over.

AI gave me two options how to simplify it.

Either do everything in jq directly:
```
# Extract, Transform, and Dedupe in one go
readarray -t REPOSITORIES < <(jq -r '
  [
    .components[].repositories[].url 
    | select(startswith("quay.io")) 
    | sub("/[^/]+$"; "/data-acceptable-bundles")
  ] 
  | unique 
  | .[]
' "${SNAPSHOT_SPEC_FILE}")

# Loop through the result
for repo in "${REPOSITORIES[@]}"; do
    echo "Processing unique repo: $repo"
done
```

Or use Bash Associative Arrays:
```
# 1. Declare an associative array (requires Bash 4.0+)
declare -A repo_set

# 2. Extract raw URLs in one pass (much faster than looping)
raw_urls=$(jq -r '.components[].repositories[].url' "${SNAPSHOT_SPEC_FILE}")

# 3. Loop through raw strings in Bash
while IFS= read -r url; do
    # Filter for quay.io
    if [[ "$url" == quay.io* ]]; then
        # Transform (using Bash string manipulation, no awk needed)
        # ${url%/*} removes the last slash and everything after it
        transformed="${url%/*}/data-acceptable-bundles"
        
        # Add to the "Set" (The value '1' doesn't matter, the Key is what we want)
        repo_set["$transformed"]=1
    fi
done <<< "$raw_urls"

# 4. Iterate over the keys (the unique URLs)
for repo in "${!repo_set[@]}"; do
    echo "Processing: $repo"
done
```

WDYT? The first one looks really simple. But of course some comments would be useful.

kasemAlem replied:
agree, 
the first option seems to be optimal one 
fixed

mmalina commented on tasks/managed/make-repo-public/tests/mocks.sh:
```diff
@@ -7,13 +7,21 @@ function curl() {
       return 1
     fi
+    # Allow data-acceptable-bundles repos and other repo patterns
+    if [[ "$*" == *data-acceptable-bundles* ]] || [[ "$*" == *redhat-services-*'/myrepo'* ]]
```
Isn't the `*redhat-services-*'/myrepo'*` pattern redundant here? Because any string that matches this will also match `*redhat-services-prod/myrepofailing*` and in such case the function will already return 1, no?

Also, I don't understand the logic. All of this new code seems redundant. The original logic was: If it's `*redhat-services-prod/myrepofailing*`, return 1, otherwise do nothing (i.e. return 0).

From looking at this, it will still behave exactly the same - fail for `*redhat-services-prod/myrepofailing*`, succeed for all other cases. So why add those lines?

kasemAlem replied:
You're right. 
The explicit checks are redundant. 
The original logic was:
   If it's `redhat-services-prod/myrepofailing` → return 1 (fail)
   Otherwise → return 0 (succeed)
The explicit return 0 statements don't change behavior. 
I'll simplify the mock back to the original logic.

mmalina replied:
ok

mmalina commented on tasks/managed/prepare-data-acceptable-bundles/tests/test-prepare-data-acceptable-bundles.yaml:
```diff
@@ -0,0 +1,270 @@
+  description: |
+    Run the prepare-data-acceptable-bundles task and verify
+    that it correctly modifies the snapshot to include data-acceptable-bundles
+    repositories as components marked as public.
```
Hmm, shouldn't it honor the `public` setting of the component? Your input data does not specify `public` at all, so I assume it will always set it to public. Is that correct?

kasemAlem replied:
yes, 
now fixed this : 

The task now honors the public setting:
1. Only components with explicit public: true are processed
2. Data-acceptable-bundles components are created only for repositories from components that are already marked as public
3. This aligns with make-repo-public, which only processes components with public: true

mmalina replied:
ok, good
